chainerx_cc/chainerx/cuda/memory_pool.h

```diff
@@ -53,13 +57,56 @@ class PinnedMemoryAllocator : public Allocator {
     void Free(void* ptr) noexcept override { cudaFreeHost(ptr); }
 };
 
+// A chunk that points to a device memory.
+//
+// A chunk might be a splitted memory block from a larger allocation.
+// The prev/next pointers construct a doubly-linked list of contiguous memories
+// sorted by those base addresses.
+class Chunk {
+public:
+    Chunk(void* ptr, size_t offset, size_t bytesize)
+        : ptr_{reinterpret_cast<void*>(reinterpret_cast<intptr_t>(ptr) + offset)}, bytesize_(bytesize) {}
```

Comment:
Maybe, `NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)` is needed here, too.